artifacts.go

```diff
@@ -301,205 +294,477 @@ func validateArtifact(c *cli.Context) error {
+			return errors.Wrapf(fileErr,
+				"can not create script file: %v", sLocation)
 		}
-		getNext = true
+		defer f.Close()
 
-		// check if we are not having element out of order
-		if elem.required && !strings.HasPrefix(raw.Name, elem.name) {
-			return errors.Errorf("Invalid artifact, should contain '%s' "+
-				"file , but contains '%s'", elem.name, raw.Name)
-		} else if !elem.required && !strings.HasPrefix(raw.Name, elem.name) {
-			// we have missing optional element; move on to the next one
-			getNext = false
-			continue
+		_, err = io.Copy(f, r)
+		if err != nil {
+			return errors.Wrapf(err,
+				"can not write script file: %v", sLocation)
 		}
+		f.Sync()
+		return nil
+	}
 
-		// check if we are having "manifest" file, which we need to sign
-		if raw.Name == "manifest" {
-			if err = createSignature(raw, w, key); err != nil {
-				return err
-			}
-			continue
+	verify := func(message, sig []byte) error {
+		return nil
+	}
 
-		} else if raw.Name == "manifest.sig" && !force {
-			// we are re-signing the artifact; return error by default
-			return errors.New("Trying to sign already signed artifact; " +
-				"please use force option")
-		} else if raw.Name == "manifest.sig" && force {
-			// just continue here as new signature is already part of tmp artifact
-			continue
+	data := dataFile
+	ar := areader.NewReader(from)
+
+	if dataFile == "" {
+		tmpData, tmpErr := ioutil.TempFile("", "mender-repack")
+		if tmpErr != nil {
+			return nil, tmpErr
 		}
+		defer os.Remove(tmpData.Name())
 
-		if err = w.WriteRaw(raw); err != nil {
-			return errors.Wrap(err, "Can not write artifact")
+		rootfs := handlers.NewRootfsInstaller()
+		rootfs.InstallHandler = func(r io.Reader, df *handlers.DataFile) error {
+			_, err = io.Copy(tmpData, r)
+			return err
 		}
+
+		data = tmpData.Name()
+		ar.RegisterHandler(rootfs)
 	}
-	return nil
+
+	r, err := read(ar, verify, storeScripts)
+	if err != nil {
+		return nil, err
+	}
+
+	info := r.GetInfo()
+
+	// now once arifact is read we need to
+	var h *handlers.Rootfs
+	switch info.Version {
+	case 1:
+		h = handlers.NewRootfsV1(data)
+	case 2:
+		h = handlers.NewRootfsV2(data)
+	default:
+		return nil, errors.Errorf("unsupported artifact version: %d", info.Version)
+	}
+
+	upd := &awriter.Updates{
+		U: []handlers.Composer{h},
+	}
+	scr, err := scripts([]string{sDir})
+	if err != nil {
+		return nil, err
+	}
+
+	aWriter := awriter.NewWriter(to)
+	if key != nil {
+		aWriter = awriter.NewWriterSigned(to, artifact.NewSigner(key))
+	}
+
+	name := ar.GetArtifactName()
+	if newName != "" {
+		name = newName
+	}
+	err = aWriter.WriteArtifact(info.Format, info.Version,
+		ar.GetCompatibleDevices(), name, upd, scr)
+
+	return ar, err
+}
+
+// oblivious to whether the file exists beforehand
+func modifyName(name, image string) error {
+	data := fmt.Sprintf("artifact_name=%s", name)
+	tmpNameFile, err := ioutil.TempFile("", "mender-name")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpNameFile.Name())
```

Comment:
Needs `tmpNameFile.Close()` in all return paths prior to natural `tmpNameFile.Close()`. See my other more general comment (search for "Go drawback").

Reply:
Fixed!